packages/next/pages/_document.tsx

```diff
@@ -253,30 +253,58 @@ export class Head extends Component<
   }
 
   getPreloadMainLinks(files: DocumentFiles): JSX.Element[] | null {
-    const { assetPrefix, devOnlyCacheBusterQueryString } = this.context
+    const {
+      assetPrefix,
+      devOnlyCacheBusterQueryString,
+      scriptLoader,
+    } = this.context
     const preloadFiles = files.allFiles.filter((file: string) => {
       // `dynamicImports` will contain both `.js` and `.module.js` when
       // the feature is enabled. This clause will filter down to the
       // modern variants only.
       return file.endsWith(getOptionalModernScriptVariant('.js'))
     })
 
-    return !preloadFiles.length
-      ? null
-      : preloadFiles.map((file: string) => (
-          <link
-            key={file}
-            nonce={this.props.nonce}
-            rel="preload"
-            href={`${assetPrefix}/_next/${encodeURI(
-              file
-            )}${devOnlyCacheBusterQueryString}`}
-            as="script"
-            crossOrigin={
-              this.props.crossOrigin || process.env.__NEXT_CROSS_ORIGIN
-            }
-          />
-        ))
+    return [
+      ...(scriptLoader.eager || []).map((file) => (
```

Comment:
Dangerously blocking scripts should be preloaded too, right?

Reply:
Dangerously blocking script options is just the fallback option, so we don't want to preload all those scripts

Reply:
Wouldn't that allow people to shoot themselves in the foot easily though? I imagine we should force a preload unless the script contents is inline.

Do we want to disallow using `async` and `defer` with the render blocking scripts and instead have them use the built-in option? Truthfully, I'm not a fan of `dangerouslyBlockRendering` — what's the use case of having this versus just telling you to use a `<script>`?

Reply:
The main idea is we now force users to consciously make a decision to make a script load in a render blocking way by selecting the `dangerouslyBlockRendering` option.
We've seen this problem earlier where a lot of script tags were included without really considering the impact.

The second benefit is it would also point to a useful usecase which will help the api evolve. As more users use `dangerouslyBlockRendering` they would complain that thier usecase in fact isn't dangerous at all which is good insight

I would be very worried to add preloads to `dangerouslyBlockRendering` since these scripts may or may not be render blocking in the first place.